akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -193,173 +264,78 @@ private[akka] class AffinityPool(
+    final val thread: Thread = threadFactory.newThread(this)
+    @volatile private[this] var executing: Boolean = false
+
+    final def start(): Unit =
+      if (thread eq null) throw new IllegalStateException(s"Was not able to allocate worker thread for ${AffinityPool.this}")
+      else thread.start()
+
+    override final def run(): Unit = {
+      // Returns true if it executed something, false otherwise
+      def executeNext(): Boolean = {
+        val c = q.poll()
+        if (c ne null) {
+          executing = true
+          try
+            c.run()
+          finally
+            executing = false
+          idleStrategy.reset()
+          true
+        } else {
+          idleStrategy.idle() // if not wait for a bit
+          false
+        }
       }
-    }
-
-    def reset(): Unit = {
-      spins = 0
-      yields = 0
-      state = NotIdle
-    }
-
-  }
-
-  private final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
-
-  private final class ThreadPoolWorker(val q: BoundedTaskQueue, val idleStrategy: IdleStrategy) extends Runnable {
-
-    private sealed trait WorkerState
-    private case object NotStarted extends WorkerState
-    private case object InExecution extends WorkerState
-    private case object Idle extends WorkerState
-
-    val thread: Thread = tf.newThread(this)
-    @volatile private var workerState: WorkerState = NotStarted
-
-    def startWorker(): Unit = {
-      workerState = Idle
-      thread.start()
-    }
-
-    private def runCommand(command: Runnable) = {
-      workerState = InExecution
-      try
-        command.run()
-      finally
-        workerState = Idle
-    }
-
-    override def run(): Unit = {
 
       /**
-       * Determines whether the worker can keep running or not.
-       * In order to continue polling for tasks three conditions
-       * need to be satisfied:
-       *
-       * 1) pool state is less than Shutting down or queue
-       * is not empty (e.g pool state is ShuttingDown but there are still messages to process)
-       *
-       * 2) the thread backing up this worker has not been interrupted
-       *
-       * 3) We are not in ShutDown state (in which we should not be processing any enqueued tasks)
+       * We keep running as long as we are Running
+       * or we're ShuttingDown but we still have tasks to execute,
+       * and we're not interrupted.
        */
-      def shouldKeepRunning =
-        (poolState < ShuttingDown || !q.isEmpty) &&
-          !Thread.interrupted() &&
-          poolState != ShutDown
+      @tailrec def runLoop(): Unit =
+        if (!Thread.interrupted()) {
+          (poolState: @switch) match {
+            case Uninitialized ⇒ ()
+            case Initializing | Running ⇒
+              executeNext()
+              runLoop()
+            case ShuttingDown ⇒
+              if (executeNext()) runLoop()
```

Comment:
@zaharidichev If you can come up with some good tests for that then it would be great! You can always set up some runnables which you can control when/if they have started to run, and when they should complete their execution. Let me know if you want/need my feedback on something like that.